tensorflow/core/kernels/segment_reduction_ops.cc

```diff
@@ -183,6 +186,54 @@ class SegmentReductionOp : public OpKernel {
   }
 };
 
+#ifdef GOOGLE_CUDA
+//  SegmentSumGPUOp is a segment sum operator implemented for GPU only.
+template <class T, class Index>
+class SegmentSumGPUOp : public OpKernel {
+ public:
+  explicit SegmentSumGPUOp(OpKernelConstruction* context) : OpKernel(context) {}
+
+  void Compute(OpKernelContext* context) override {
+    const Tensor& input = context->input(0);
+    const Tensor& segment_ids = context->input(1);
+
+    if (!SegmentReductionDoValidation(context, input, segment_ids)) {
+      return;
+    }
+
+    const int64 num_indices = segment_ids.NumElements();
+
+    Index output_rows;
+    context->eigen_device<GPUDevice>().memcpyDeviceToHost(
```

Comment:
This introduces a device/host sync. TensorFlow tries very hard to avoid that. 

An example on how to address such issue: 

https://github.com/tensorflow/tensorflow/blob/master/tensorflow/core/kernels/where_op.cc#L253

Another common practice is to check the indices on GPU, and do nothing with invalid indicies.